File: Codility/04. Counting Elements/MissingInteger/Solution.swift
// https://app.codility.com/programmers/lessons/4-counting_elements/missing_integer/
// Respectable

public func solution(_ A : inout [Int]) -> Int {
    let N = A.count
    var M = [Int](repeating: 0, count: 100001)
    
    for I in 0..<N {
        if A[I] <= 0 {
            continue
        }
        
        M[A[I] - 1] += 1
    }
    
    for I in 0..<M.count {
        if M[I] == 0 {
            return I + 1
        }
    }
    
    return 1
}

File: Codility/04. Counting Elements/MissingInteger/Solution.py
# https://app.codility.com/programmers/lessons/4-counting_elements/missing_integer/
# Respectable

from typing import List


def solution(A: List[int]) -> int:
    """Build a lookup set, then scan upward from 1 for the first integer not present."""
    # Time: O(n)   Space: O(n)
    seen = set(A)
    candidate = 1
    while candidate in seen:
        candidate += 1
    return candidate


def test() -> None:
    assert solution([1, 3, 6, 4, 1, 2]) == 5
    assert solution([1, 2, 3]) == 4
    assert solution([-1, -3]) == 1
    assert solution([]) == 1
